tcms/telemetry/static/telemetry/js/testing/status-matrix.js

```diff
@@ -0,0 +1,116 @@
+let table;
+let initial_column = {
+    data: null,
+    className: "table-view-pf-actions",
+    render: function (data, type, full, meta) {
+        const caseId = data.tc_id;
+
+        return `<span style="padding: 5px;">` +
+            `<a href="/case/${caseId}">TC-${caseId}: ${data.tc_summary}</a>` +
+            `</span>`;
+    }
+};
+
+$(document).ready(() => {
+    loadInitialProduct();
+    loadInitialTestPlans();
+
+    document.getElementById('id_product').onchange = () => {
+        update_version_select_from_product();
+        update_build_select_from_product(true);
+        updateTestPlanSelectFromProduct(drawTable);
+    };
+
+    document.getElementById('id_version').onchange = drawTable;
+    document.getElementById('id_build').onchange = drawTable;
+    document.getElementById('id_test_plan').onchange = drawTable;
+
+    $('#id_after').on('dp.change', drawTable);
+    $('#id_before').on('dp.change', drawTable);
+
+    drawTable();
+});
+
+
+function drawTable() {
+    if (table) {
+        table.destroy();
+
+        $('table > thead > tr > th:not(.header)').remove();
+        $('table > tbody > tr').remove();
+    }
+
+    const query = {};
+
+    const productId = $('#id_product').val();
+    if (productId) {
+        query['case__plan__product'] = productId;
+    }
+
+    const versionId = $('#id_version').val();
+    if (versionId) {
+        query['case__plan__product_version'] = versionId;
+    }
+
+    const buildId = $('#id_build').val();
+    if (buildId) {
+        query['build_id'] = buildId;
+    }
+
+    const testPlanId = $('#id_test_plan').val();
+    if (testPlanId) {
+        query['case__plan__plan_id'] = testPlanId;
+    }
+
+    const dateBefore = $('#id_before');
+    if (dateBefore.val()) {
+        query['close_date__lte'] = dateBefore.data('DateTimePicker').date().format('YYYY-MM-DD 23:59:59');
+    }
+
+    const dateAfter = $('#id_after');
+    if (dateAfter.val()) {
+        query['close_date__gte'] = dateAfter.data('DateTimePicker').date().format('YYYY-MM-DD 00:00:00');
+    }
+
+    jsonRPC('Testing.status_matrix', query, data => {
+
+        const table_columns = [initial_column];
+
+        Object.keys(data.columns).forEach(testRunId => {
+            const testRunSummary = data.columns[testRunId];
+            $('.table > thead > tr').append(`<th>TR-${testRunId} ${testRunSummary}</th>`);
+
+            table_columns.push({
+                data: null,
+                sortable: false,
+                render: renderData(testRunId)
+            });
+        });
+
+        table = $('#table').DataTable({
+            columns: table_columns,
+            data: data.data,
+            paging: false,
+            ordering: false,
+            dom: "t"
+        });
+
+        ['idle', 'pass', 'fail', 'other'].forEach(className => {
```

Comment:
This is duplicating the status class names inside the JS code which will make it yet another place to update when we decide to change these names. Better use 2 class names, e.g. 'execution-status pass'. Then filter via the 'execution-status' class which is only used for the selectors. The actual color will still come from the other class name. As long as you apply both to the parent element everything should be fine and can drop the other loop.

Reply:
Done. You are right that this is not and exposes some internal properties. Unfortunately, changing it meant I had to add one more quite long and ugly js function, because of the way the DOM API works.